packages/typegpu/src/tgpuBindGroupLayout.ts

```diff
@@ -0,0 +1,452 @@
+   */
+  unwrap(unwrapper: Unwrapper): GPUBindGroupLayout;
+}
+
+type StorageUsageForEntry<T extends TgpuLayoutStorage> = T extends {
+  access?: infer Access;
+} // Is the access defined on the type?
+  ? 'mutable' | 'readonly' extends Access // Is the access ambiguous?
+    ? TgpuBufferReadonly<T['storage']> | TgpuBufferMutable<T['storage']>
+    : 'readonly' extends Access // Is the access strictly 'readonly'?
+      ? TgpuBufferReadonly<T['storage']>
+      : 'mutable' extends Access // Is the access strictly 'mutable'?
+        ? TgpuBufferMutable<T['storage']>
+        : TgpuBufferReadonly<T['storage']> | TgpuBufferMutable<T['storage']>
+  : TgpuBufferReadonly<T['storage']>; // <- access is undefined, so default to 'readonly';
+
+export type LayoutEntryToInput<T extends TgpuLayoutEntry | null> =
+  T extends TgpuLayoutUniform
+    ?
+        | TgpuBufferUsage<T['uniform'], 'uniform'>
+        | (TgpuBuffer<T['uniform']> & Uniform)
+        | GPUBuffer
+    : T extends TgpuLayoutStorage
+      ?
+          | StorageUsageForEntry<T>
+          | (TgpuBuffer<T['storage']> & Storage)
+          | GPUBuffer
+      : T extends TgpuLayoutSampler
+        ? GPUSampler
+        : T extends TgpuLayoutTexture
+          ? GPUTextureView
+          : T extends TgpuLayoutStorageTexture
+            ? GPUTextureView
+            : T extends TgpuLayoutExternalTexture
+              ? GPUExternalTexture
+              : never;
+
+export type BindLayoutEntry<T extends TgpuLayoutEntry | null> =
+  T extends TgpuLayoutUniform
+    ? TgpuBufferUniform<T['uniform']>
+    : T extends TgpuLayoutStorage
+      ? StorageUsageForEntry<T>
+      : T extends TgpuLayoutSampler
+        ? TgpuSampler
+        : never;
+
+export type TgpuBindGroup<
+  Entries extends Record<string, TgpuLayoutEntry | null> = Record<
+    string,
+    TgpuLayoutEntry | null
+  >,
+> = {
+  readonly resourceType: 'bind-group';
+  readonly layout: TgpuBindGroupLayout<Entries>;
+  unwrap(unwrapper: Unwrapper): GPUBindGroup;
+};
+
+export function bindGroupLayout<
+  Entries extends Record<string, TgpuLayoutEntry | null>,
+>(entries: Entries): TgpuBindGroupLayout<Entries> {
+  return createBindGroupLayout(entries);
+}
+
+export function isBindGroupLayout<T extends TgpuBindGroupLayout>(
+  value: T | unknown,
+): value is T {
+  return !!value && (value as T).resourceType === 'bind-group-layout';
+}
+
+export function isBindGroup<T extends TgpuBindGroup>(
+  value: T | unknown,
+): value is T {
+  return !!value && (value as T).resourceType === 'bind-group';
+}
+
+/**
+ * @category Errors
+ */
+export class MissingBindingError extends Error {
+  constructor(groupLabel: string | undefined, key: string) {
+    super(
+      `Bind group '${groupLabel ?? '<unnamed>'}' is missing a required binding '${key}'`,
+    );
+
+    // Set the prototype explicitly.
+    Object.setPrototypeOf(this, MissingBindingError.prototype);
+  }
+}
```

Comment:
Maybe move this to `errors`?

Reply:
It was in errors at first, but I thought it would make more sense to keep this local to here, as it strictly relates to bind group layouts.